File: debop4s-timeperiod/src/test/scala/com/github/debop4s/timeperiod/tests/calendars/seekers/DaySeekerTest.scala
package com.github.debop4s.timeperiod.tests.calendars.seekers

import com.github.debop4s.timeperiod._
import com.github.debop4s.timeperiod.calendars.CalendarVisitorFilter
import com.github.debop4s.timeperiod.calendars.seeker.DaySeeker
import com.github.debop4s.timeperiod.tests.AbstractTimePeriodTest
import com.github.debop4s.timeperiod.timerange.{DayRangeCollection, DayRange}
import com.github.debop4s.timeperiod.utils.Times

/**
 * DaySeekerTest
 * Created by debop on 2014. 2. 18.
 */
class DaySeekerTest extends AbstractTimePeriodTest {

  test("simple forward") {
    val start = DayRange()
    val daySeeker = DaySeeker()

    val day1 = daySeeker.findDay(start, 0)
    assert(day1 != null)
    day1.isSamePeriod(start) should equal(true)

    val day2 = daySeeker.findDay(start, 1)
    assert(day2 != null)
    day2.isSamePeriod(start.nextDay) should equal(true)

    (-10 until 20).par.foreach { i =>
      val offset = i * 5
      val day = daySeeker.findDay(start, offset)
      day.isSamePeriod(start.addDays(offset)) should equal(true)
    }
  }

  test("simple backward") {
    val start = DayRange()
    val daySeeker = DaySeeker(SeekDirection.Backward)

    val day1 = daySeeker.findDay(start, 0)
    day1.isSamePeriod(start) should equal(true)

    (-10 until 20).par.foreach { i =>
      val offset = i * 5
      val day = daySeeker.findDay(start, offset)
      day.isSamePeriod(start.addDays(-offset)) should equal(true)
    }
  }

  test("seek direction") {
    val start = DayRange()
    val daySeeker = DaySeeker()

    (-10 until 20).par.foreach { i =>
      val offset = i * 5
      val day = daySeeker.findDay(start, offset)
      day.isSamePeriod(start.addDays(offset)) should equal(true)
    }

    val backwardSeeker = DaySeeker(SeekDirection.Backward)

    (-10 until 20).foreach { i =>
      val offset = i * 5
      val day = backwardSeeker.findDay(start, offset)
      day.isSamePeriod(start.addDays(-offset)) should equal(true)
    }
  }

  test("min date") {
    val daySeeker = DaySeeker()
    val day = daySeeker.findDay(DayRange(MinPeriodTime), -10)
    day should equal(null)
  }

  test("max date") {
    val daySeeker = DaySeeker()
    val day = daySeeker.findDay(DayRange(MaxPeriodTime), 10)
    day should equal(null)
  }

  test("seek weekend holiday") {
    val start = DayRange(Times.asDate(2011, 2, 15))

    val filter = new CalendarVisitorFilter()
    filter.addWorkingWeekdays()
    filter.excludePeriods.add(DayRangeCollection(2011, 2, 27, 14)) // 14 days => week 9 and week 10

    val daySeeker = new DaySeeker(filter)

    val day1 = daySeeker.findDay(start, 3)
    day1 should equal(DayRange(2011, 2, 18))

    // 주말 (19, 20) 제외
    daySeeker.findDay(start, 4) should equal(DayRange(2011, 2, 21))

    // 주말 (19, 20) 제외, 2.27 일부터 14일간 휴가
    daySeeker.findDay(start, 10) should equal(DayRange(2011, 3, 15))
  }
}

File: debop4s-timeperiod/src/test/java/debop4s/timeperiod/tests/calendars/seekers/DaySeekerTest.java
package debop4s.timeperiod.tests.calendars.seekers;

import debop4s.core.JAction1;
import debop4s.core.parallels.JParallels;
import debop4s.timeperiod.SeekDirection;
import debop4s.timeperiod.TimeSpec;
import debop4s.timeperiod.calendars.CalendarVisitorFilter;
import debop4s.timeperiod.calendars.seeker.DaySeeker;
import debop4s.timeperiod.tests.TimePeriodTestBase;
import debop4s.timeperiod.timerange.DayRange;
import debop4s.timeperiod.timerange.DayRangeCollection;
import debop4s.timeperiod.utils.Times;
import lombok.extern.slf4j.Slf4j;
import org.junit.Test;

import static org.fest.assertions.Assertions.assertThat;

/**
 * debop4s.timeperiod.test.calendars.seekers.DaySeekerFunSuite
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 13. 5. 27. 오후 6:16
 */
@Slf4j
public class DaySeekerTest extends TimePeriodTestBase {

    @Test
    public void simpleForward() {

        final DayRange start = new DayRange();
        final DaySeeker daySeeker = new DaySeeker();

        DayRange day1 = daySeeker.findDay(start, 0);
        assertThat(day1.isSamePeriod(start)).isTrue();

        DayRange day2 = daySeeker.findDay(start, 1);
        assertThat(day2.isSamePeriod(start.nextDay())).isTrue();

        JParallels.run(-10, 20, new JAction1<Integer>() {
            @Override
            public void perform(Integer i) {
                Integer offset = i * 5;
                DayRange day = daySeeker.findDay(start, offset);
                assertThat(day.isSamePeriod(start.addDays(offset))).isTrue();
            }
        });
    }

    @Test
    public void simpleBackward() {

        final DayRange start = new DayRange();
        final DaySeeker daySeeker = new DaySeeker(SeekDirection.Backward);

        DayRange day1 = daySeeker.findDay(start, 0);
        assertThat(day1.isSamePeriod(start)).isTrue();

        DayRange day2 = daySeeker.findDay(start, 1);
        assertThat(day2.isSamePeriod(start.previousDay())).isTrue();

        JParallels.run(-10, 20, new JAction1<Integer>() {
            @Override
            public void perform(Integer i) {
                Integer offset = i * 5;
                DayRange day = daySeeker.findDay(start, offset);
                assertThat(day.isSamePeriod(start.addDays(-offset))).isTrue();
            }
        });
    }

    @Test
    public void seekDirectionTest() {
        final DayRange start = new DayRange();
        final DaySeeker daySeeker = new DaySeeker();

        JParallels.run(-10, 20, new JAction1<Integer>() {
            @Override
            public void perform(Integer i) {
                Integer offset = i * 5;
                DayRange day = daySeeker.findDay(start, offset);
                assertThat(day.isSamePeriod(start.addDays(offset))).isTrue();
            }
        });

        final DaySeeker backwardSeeker = new DaySeeker(SeekDirection.Backward);

        JParallels.run(-10, 20, new JAction1<Integer>() {
            @Override
            public void perform(Integer i) {
                Integer offset = i * 5;
                DayRange day = backwardSeeker.findDay(start, offset);
                assertThat(day.isSamePeriod(start.addDays(-offset))).isTrue();
            }
        });
    }

    @Test
    public void minDateTest() {
        DaySeeker daySeeker = new DaySeeker();
        DayRange day = daySeeker.findDay(new DayRange(TimeSpec.MinPeriodTime), -10);
        assertThat(day).isNull();
    }

    @Test
    public void maxDateTest() {
        DaySeeker daySeeker = new DaySeeker();
        DayRange day = daySeeker.findDay(new DayRange(TimeSpec.MaxPeriodTime), 10);
        assertThat(day).isNull();
    }

    @Test
    public void seekWeekendHolidayTest() {

        DayRange start = new DayRange(Times.asDate(2011, 2, 15));

        CalendarVisitorFilter filter = new CalendarVisitorFilter();
        filter.addWorkingWeekdays();
        filter.getExcludePeriods().add(new DayRangeCollection(2011, 2, 27, 14)); // 14 days -> week 9 and week 10

        DaySeeker daySeeker = new DaySeeker(filter);

        DayRange day1 = daySeeker.findDay(start, 3);
        assertThat(day1).isEqualTo(new DayRange(2011, 2, 18));

        DayRange day2 = daySeeker.findDay(start, 4);                // 주말 (19, 20) 제외
        assertThat(day2).isEqualTo(new DayRange(2011, 2, 21));

        DayRange day3 = daySeeker.findDay(start, 10);                // 주말 (19, 20) 제외, 2.27부터 14일간 휴가
        assertThat(day3).isEqualTo(new DayRange(2011, 3, 15));
    }

}
